internal/plugin/external_plugins.go

```diff
@@ -0,0 +1,228 @@
+package plugin
+
+import (
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	go_plugin "plugin"
+	"strings"
+
+	"github.com/cyberark/secretless-broker/pkg/secretless/log"
+	"github.com/cyberark/secretless-broker/pkg/secretless/plugin/connector/http"
+	"github.com/cyberark/secretless-broker/pkg/secretless/plugin/connector/tcp"
+)
+
+// ExternalPluginLookupFunc returns all available external plugins.
+type ExternalPluginLookupFunc func(pluginDir string,
+	checksumfile string,
+	logger log.Logger,
+) (map[string]*go_plugin.Plugin, error)
+
+// LoadPluginsFromDir loads all plugins from a specified directory and returns
+// Plugins struct with tcp and http connectors.
+func LoadPluginsFromDir(pluginDir string,
+	checksumsFile string,
+	logger log.Logger,
+) (map[string]*go_plugin.Plugin, error) {
+
+	filenames, err := getCheckedPlugins(pluginDir, checksumsFile, logger)
+	if err != nil {
+		return nil, err
+	}
+
+	return loadPluginFiles(pluginDir, filenames, logger)
+}
+
+// ExternalPlugins is used to enumerate all externally-available plugins in a sepcified
+// directory to the clients of this method.
+//TODO: Test this
+func ExternalPlugins(
+	pluginDir string,
+	lookupFunc ExternalPluginLookupFunc,
+	logger log.Logger,
+	checksumsFile string,
+) (AvailablePlugins, error) {
+
+	rawPlugins, err := lookupFunc(pluginDir, checksumsFile, logger)
+	if err != nil {
+		return nil, err
+	}
+
+	plugins := Plugins{
+		HTTPPluginMap: map[string]http.Plugin{},
+		TCPPluginMap:  map[string]tcp.Plugin{},
+	}
+
+	for rawPluginName, rawPlugin := range rawPlugins {
+		logger.Infof("Loading plugin '%s'...", rawPluginName)
+
+		pluginType, pluginID, err := parsePluginMetadata(rawPlugin, rawPluginName, logger)
```

Comment:
the `logger` is passed to parsePluginMetadata, getSymbolFromName etc. would it not be better for them to propagate errors upwards so that they can be logged here ? as is being done below